### PR TITLE
Sofware renderer: keep a Weak of the Window

### LINKED ISSUE #1545: Remove the `window` parameter from `SoftwareRenderer`'s `render` functions
In the API review we decided that we can simplify the API of these two functions and make the more robust by

 - Letting the `SoftwareRenderer`'s `new()` function take a `&Weak<dyn WindowAdapter>` and clone it internally.
 - Remove the `window` parameter from the `render` functions.

### PR DESCRIPTION
**Pull request by ogoffart**
Closes #1545

Not that `SoftwareRenderer::new()`  takes the Weak by value instead of
by reference, because coercing a reference to a `Weak` to a `&Weak<dyn>`
is not possible, while converting a Weak to a `Weak<dyn>` just works.

### CODE REVIEW

File: internal/core/swrenderer/draw_functions.rs [hunk 1]
```diff
@@ -283,6 +283,7 @@ fn interpolate_color(
 /// PremultipliedRgbaColor can be constructed from a [`Color`] with
 /// the [`From`] trait. This conversion will pre-multiply the color
 /// components
+#[allow(missing_docs)]
```
Comment (tronical) [inline internal/core/swrenderer/draw_functions.rs:286]: Looks unrelated ;)

**Review by tronical — APPROVED** [review]

File: internal/core/swrenderer/draw_functions.rs (hunk 1 above)
Comment (ogoffart) [reply internal/core/swrenderer/draw_functions.rs:286]: I realized the new function didn't have docs, so i added the `warn`  ,  but then it warns about undocumented fields in this struct, which i think is fine.
